src/linker/Linker.Steps/SealerStep.cs

```diff
@@ -0,0 +1,148 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Collections.Generic;
+using Mono.Cecil;
+
+namespace Mono.Linker.Steps
+{
+	public class SealerStep : BaseStep
+	{
+		HashSet<TypeDefinition> referencedBaseTypeCache;
+
+		public SealerStep ()
+		{
+		}
+
+		protected override void ProcessAssembly (AssemblyDefinition assembly)
+		{
+			if (Annotations.GetAction (assembly) != AssemblyAction.Link)
+				return;
+
+			if (!Context.Optimizations.IsEnabled (CodeOptimizations.Sealer, assembly))
+				return;
+
+			foreach (var type in assembly.MainModule.Types)
+				ProcessType (type);
+		}
+
+		protected override void EndProcess ()
+		{
+			referencedBaseTypeCache = null;
+		}
+
+		bool IsSubclassed (TypeDefinition type)
+		{
+			if (referencedBaseTypeCache == null) {
+				referencedBaseTypeCache = new HashSet<TypeDefinition> ();
+				foreach (var a in Context.GetAssemblies ()) {
+					foreach (var s in a.MainModule.Types) {
+						var btd = s.BaseType?.Resolve ();
+						if (btd != null)
+							referencedBaseTypeCache.Add (btd);
+
+						if (s.HasNestedTypes) {
+							foreach (var ns in s.NestedTypes) {
+								btd = s.BaseType?.Resolve ();
+								if (btd != null)
+									referencedBaseTypeCache.Add (btd);
+							}
+						}
+					}
+				}
+			}
+
+			var bt = type.Resolve ();
+			return referencedBaseTypeCache.Contains (bt);
+		}
+
+		void ProcessType (TypeDefinition type)
+		{
+			//
+			// interface members are virtual (and we cannot change this)
+			//
+			if (type.IsInterface)
+				return;
+
+			//
+			// the code does not include any subclass for this type
+			//
+			if (!type.IsAbstract && !type.IsSealed && !IsSubclassed (type))
+				SealType (type);
+
+			if (type.HasNestedTypes) {
```

Comment:
Maybe move this all the way up so that we don't miss classes nested under interfaces?